src/consumer_queue.h

```diff
@@ -2,60 +2,167 @@
 
 	/// Destructor. Unregisters from the send buffer, if any.
 	~consumer_queue();
 
-	///  Push a new sample onto the queue.
-	void push_sample(const sample_p &sample);
+	/**
+	 * Push a new sample onto the queue. Can only be called by one thread (single-producer).
+	 * This deletes the oldest sample if the max capacity is exceeded.
+	 */
+	template<class T>
+	void push_sample(T&& sample) {
+		while (!try_push(std::forward<T>(sample))) {
+			// buffer full, drop oldest sample
+			if (!done_sync_.load(std::memory_order_acquire)) {
+				// synchronizes-with store to done_sync_ in ctor
+				std::atomic_thread_fence(std::memory_order_acquire);
+				done_sync_.store(true, std::memory_order_release);
+			}
+			try_pop();
+		}
+		{
+			// ensure that notify_one doesn't happen in between try_pop and wait_for
+			std::lock_guard<std::mutex> lk(mut_);
+			cv_.notify_one();
+		}
+	}
 
 	/**
-	 * Pop a sample from the queue.
-	 * Blocks if empty.
-	 * @param timeout Timeout for the blocking, in seconds. If expired, an empty sample is returned.
+	* Pop a sample from the queue. Can be called by multiple threads (multi-consumer).
+	* Blocks if empty and if a nonzero timeout is used.
+	* @param timeout Timeout for the blocking, in seconds. If expired, an empty sample is returned.
 	 */
-	sample_p pop_sample(double timeout = FOREVER);
+	sample_p pop_sample(double timeout = FOREVER) {
+		sample_p result;
+		bool success = try_pop(result);
+		if (!success && timeout > 0.0) {
+			// only acquire mutex if we have to do a blocking wait with timeout
+			std::chrono::duration<double> sec(timeout);
+			std::unique_lock<std::mutex> lk(mut_);
+			if (!try_pop(result))
+				cv_.wait_for(lk, sec, [&]{ return this->try_pop(result); });
+		}
+		return result;
+	}
 
-	/// Number of available samples
-	std::size_t read_available() const { return buffer_.read_available(); }
+	/// Number of available samples. This is approximate unless called by the thread calling the
+	/// pop_sample().
+	std::size_t read_available() const;
 
-	/// Flush the queue, return the number of dropped samples
+	/// Flush the queue, return the number of dropped samples.
 	uint32_t flush() noexcept;
 
-	/// Check whether the buffer is empty.
-	bool empty();
+	/// Check whether the buffer is empty. This is approximate unless called by the thread calling
+	/// the pop_sample().
+	bool empty() const;
 
 	consumer_queue(const consumer_queue&) = delete;
+	consumer_queue(consumer_queue&&) = delete;
 	consumer_queue& operator=(const consumer_queue&) = delete;
-
+	consumer_queue& operator= (consumer_queue&&) = delete;
 private:
-	send_buffer_p registry_; // optional consumer registry
-	buffer_type buffer_;	 // the sample buffer
-	std::mutex mut_;			 // mutex for cond var (also to protect queue at buffer overflow)
-	std::condition_variable cv_; // to allow for blocking wait by consumer
+	// an item stored in the queue
+	struct item_t {
+		std::atomic<std::size_t> seq_state;
```

Comment:
I haven't looked this though in detail, but I'll hopefully get to it after the 1.14 release.

Reply:
Sounds good. I hope to find some time to contrib some unit tests that stress concurrency a bit more.

Reply:
I already have some benchmarks in the unit tests that stress the queue a fair bit, but more are always welcome.